examples/cifar10_resnet.py

```diff
@@ -69,87 +100,271 @@
-    num_blocks = 3
-
-# Instantiate convolutional base (stack of blocks).
-for i in range(num_blocks):
-    for j in range(num_sub_blocks):
-        strides = 1
-        is_first_layer_but_not_first_block = j == 0 and i > 0
-        if is_first_layer_but_not_first_block:
-            strides = 2
-        y = Conv2D(num_filters,
-                   kernel_size=3,
-                   padding='same',
-                   strides=strides,
-                   kernel_initializer='he_normal',
-                   kernel_regularizer=l2(1e-4))(x)
-        y = BatchNormalization()(y)
-        y = Activation('relu')(y)
-        y = Conv2D(num_filters,
-                   kernel_size=3,
-                   padding='same',
-                   kernel_initializer='he_normal',
-                   kernel_regularizer=l2(1e-4))(y)
-        y = BatchNormalization()(y)
-        if is_first_layer_but_not_first_block:
-            x = Conv2D(num_filters,
-                       kernel_size=1,
-                       padding='same',
-                       strides=2,
-                       kernel_initializer='he_normal',
-                       kernel_regularizer=l2(1e-4))(x)
-        x = keras.layers.add([x, y])
-        x = Activation('relu')(x)
-
-    num_filters = 2 * num_filters
-
-# Add classifier on top.
-x = AveragePooling2D()(x)
-y = Flatten()(x)
-outputs = Dense(num_classes,
-                activation='softmax',
-                kernel_initializer='he_normal')(y)
-
-# Instantiate and compile model.
-model = Model(inputs=inputs, outputs=outputs)
+
+def lr_schedule(epoch):
+    """Learning Rate Schedule
+
+    Learning rate is scheduled to be reduced after 80, 120, 160, 180 epochs.
+    Called automatically every epoch as part of callbacks during training.
+
+    # Arguments
+        epoch (int): The number of epochs
+
+    # Returns
+        lr (float32): learning rate
+    """
+    lr = 1e-3
+    if epoch > 180:
+        lr *= 0.5e-3
+    elif epoch > 160:
+        lr *= 1e-3
+    elif epoch > 120:
+        lr *= 1e-2
+    elif epoch > 80:
+        lr *= 1e-1
+    print('Learning rate: ', lr)
+    return lr
+
+
+def conv_bn(inputs, num_filters=16, kernel_size=3, strides=1):
+    """2D Convolution-Batch Normalization stack builder
+
+    # Arguments
+        inputs (tensor): input tensor from input image or previous layer
+        num_filters (int): Conv2D number of filters
+        kernel_size (int): Conv2D square kernel dimensions
+        strides (int): Conv2D square stride dimensions
+
+    # Returns
+        x (tensor): tensor as input to the next layer
+    """
+    x = Conv2D(num_filters,
+               kernel_size=kernel_size,
+               strides=strides,
+               padding='same',
+               kernel_initializer='he_normal',
+               kernel_regularizer=l2(1e-4))(inputs)
+    x = BatchNormalization()(x)
+    return x
+
+
+def conv_bn_relu(inputs, num_filters=16, kernel_size=3, strides=1):
```

Comment:
Why not have an `activation` argument in `conv_bn` instead? Seems better than hard-coding `relu` *as a function*. This additional function doesn't seem necessary (same for `bn_relu_conv` -- you could simply have an argument in `conv_bn` to determine where the activation goes).

This is important in particular to make it easy to change the activation easily, and make model easier to parametrize (e.g. if the activation/bn/conv order is a function argument, that's easier to change globally than if it's a function name).

Reply:
parameterized resnet_block() function to replace all 3 functions (conv_bn_relu, conv_bn, bn_relu_conv).